### ISSUE TITLE
Asm.js, tables and indirect calls

### ISSUE THREAD

**Issue by caspervonb** [WebAssembly/binaryen#1256]
From what I can tell, an indirect call generates a call to `FUNCTION_TABLE_<SIGNATURE>`, as long as everything is the same module these work out fine.

This is where I ran into my first issue, having an indirect call triggers a parsing error if there is no table import, even if there is a table declared in the textual wasm.

In other words, the following example won't compile:
```lisp
(module
  (type (;0;) (func (param i32)))
  (type (;1;) (func))
  (func $call (type 0) (param i32)
    (call_indirect 1
      (get_local 0)))
  (func $main (type 1))
  (table (;0;) 0 anyfunc)
  (memory (;0;) 0)
  (export "call" (func $call))
  (export "main" (func $main)))
```

If there are no "direct" indirect calls, e.g trampolining to JavaScript glue then there's no error and both imports and exports compile but do not generate anything useful.

The import just becomes an unused variable, and the export tries to export `$00`.

I also can't see any correlation between `FUNCTION_TABLE_*` indices and the indirect indices?

TL;DR; Can wasm2asm support function tables with the same indexing semantics as WebAssembly?

**Comment by kripken** [comment]
Looks like the parse error on that wast is because the table is defined after the function that depends on it. Perhaps we should make the s-parser look ahead for the table definition, but it currently doesn't.

Yeah, wasm and asm.js semantics are different for indirect calls, so it's not trivial to connect them. One way would be to replace `call_indirect` with a call to a helper function that picks the right asm.js table based on the index being used. Another option is to standardize on N arguments of type double, and convert to that for all indirect calls (so just one asm.js table). Maybe a better option but I don't see one right now.

Btw, what are you doing here, for context?

**Comment by caspervonb** [comment]
> Perhaps we should make the s-parser look ahead for the table definition, but it currently doesn't.

Noted, wasm2wat issue again, it has different ordering than wasm-dis.
Mental note; Use binaryen tools for binaryen tools.

> Yeah, wasm and asm.js semantics are different for indirect calls,

Is that actual Asm.js semantics or Emscripten semantics that can be thrown away?

From what I can tell a valid table in WebAssembly pretty much tells us the assignment operations for the elements? 

e.g, a module with a table always looks something like the following.

```wat
(module 
  ;; ...
  (table (;0;) 1 anyfunc)
  (elem (i32.const 0) $func))
  ;; ...
)
```

From this it seems trivial to generate tables with the same indexing

```javascript
function asm(global, env, buffer) { 
  // (table (;0;) 1 anyfunc)
  var table = {
    elements: new Array(1),
    get: function(index) {
      index = index | 0;

      return this.elements[index];
    },
    set: function(index, value) {
      index = index | 0;
      this.elements[index] = value;
    },
    grow: function(size) {
      // throw, resize array?
    },
  };

  // (elem (i32.const 0) $func))
  table.set(0 | 0, func);

  return {
  };
```

This first pass doesn't touch `call_indirect`, so it can live next to the current semantics.
Why? it will allow tables to work somewhat like what one would expect from a WebAssembly module and so "host" code/libraries do not need to need to know if they are working with a WebAssembly.Table or ad-hoc Asm.js table.

> Btw, what are you doing here, for context?

TL;DR; I'm passing function pointers to JavaScript.

That example above was just a minimal reproduction, did a bunch of variations to see where it was failing and also trying to figure out how the table's were indexed.

Actual use case is calling function pointers from JavaScript in some minimal platform glue, it expects `table` to be available at some point, import or export doesn't matter.

```javascript
foreign['platform_schedule_tick_callback'] = function(targetId, callbackId, userOffset) {
  var target = document.getElementById(targetId);
  var callback = foreign['table'].get(callbackId);

  window.requestAnimationFrame(function(timestamp) {
    var result = callback(targetId, timestamp, userOffset);
    // et cetera
  }, target);
};
```

Asm.js comes in as a fallback for browsers that do not support WebAssembly, which is still quite a few especially on low end mobile.

**Comment by caspervonb** [comment]
Uhm yeah just realized the above is not legal Asm.js, with objects and all, hmph.

**Comment by binji** [comment]
> Noted, wasm2wat issue again, it has different ordering than wasm-dis.

Yeah, wasm2wat writes everything in the order found in the binary file, for consistency. The text format allows for nearly everything to be in any order, with the exception of imports, which must come before all other functions, globals, memories, or tables.

**Comment by kripken** [comment]
> Is that actual Asm.js semantics or Emscripten semantics that can be thrown away?

Yes, in asm.js each table has a specific function type. So one table for `void (int, float)` functions, etc. etc. All the functions in that table have that type, and all calls to that table provide the right types in the arguments. Whereas in wasm, a table can contain a mix of function types, and each call defines the argument types, which are checked at runtime if they match the function that is called from the table.

So the difficulty here is to create multiple asm.js tables from a single wasm table.

**Comment by tlively** [comment]
Closing because we now support the standard text format, so this should be fixed.
